### PR TITLE
v1.5.0

### PR DESCRIPTION
**Pull request by JoryHogeveen**
**Ready for testing!**

This PR will address version 1.5.0, fixing mutiple issues:

## Changelog

- **Feature:** Option to try and fix the widget parameters if they are invalid: #24 
- **Feature:** Option to remove duplicate classes: #25 
  - This method also enhances the speed of how classes are added!
    - _In a loop of 500 times the new method is approx 3 times faster.
In a loop of 50000 times over 30 times!_
- **Enhancement:** Sort classes based on the predefined classes on the frontend by default. #19
- **Enhancement:** Classes filter for frontend (for sorting or modifications): #19 
  - `widget_css_classes` modify all classes added by this plugin.
  - `widget_css_classes_custom` modify custom input classes.
- **Enhancement:** Plugin settings filter (`widget_css_classes_set_settings`), overwrites user settings: #16 
- **Enhancement:** Plugin default settings filter (`widget_css_classes_default_settings`): #4 
- **Enhancement:** Capability filters for form fields: #21 
  - `widget_css_classes_id_input_capability` ID input
  - `widget_css_classes_class_input_capability` classes input
  - `widget_css_classes_class_select_capability` predefined classes select (also hides classes input if invalid)
- **Compatibility:** WP External Links: #17, thanks to Victor [@freelancephp](https://profiles.wordpress.org/freelancephp)
- **Fix:** Form wrapper div style: #18, thanks to Chuck Reynolds [@ryno267](https://profiles.wordpress.org/ryno267)
- **Fix:** Enable sortable input selection (IE-11 fix): #20
- **UI:** Enhance setting page JavaScript and remove `relCopy` library dependency.
- **i18n:** Remove Dutch and Russian languages from plugin distribution (available on [translate.wordpress.org](https://translate.wordpress.org/projects/wp-plugins/widget-css-classes)). #23
- CodeClimate fixes (related PR: #15)

## Core changes:
- Refactoring for CC fixes
- Added setting handlers in the library to make sure settings are always of the correct type and format
  - Also added a unit test for this.
- Simplified code where possible
- Improved inline PHPdocs

## Tests:
**Settingpage & Widget form**
- [x] Check saving and displaying options.
- [x] Try to save wrong stuff by manually edit the form content in the console.
  - [x] Add some malicious stuff in there.

**Widget form**
- [x] Test capability filters
  - [x] Existing settings should save even if the input isn't shown
  - [x] Disallowing predefined select should also hide the custom CSS input

**Frontend**
- [x] Check correct rendering of all set classes (custom, predefined and other)
- [x] Check order of predefined classes
- [x] Check params fix option (register sidebar with empty before and after parameters to test)
- [x] Check appending classes with double and single quote format attributes.
- [x] Removing duplicates

**ImportExport**
- [x] Import from current version after fresh export
- [x] Import from older versions (see update method in library)
  - [x] 1.1
  - [x] 1.2
  - [x] 1.4

**Updating**
- [x] Configure plugin in an older version and update to the new version. All settings should stay the same and the front-end should work without changes (excerpt for class order)
    - [x] 1.1
    - [x] 1.2
    - [x] 1.4

### CODE REVIEW

**Comment by JoryHogeveen** [comment]
@cleverness 
Let me know if you see anything you don't like or you would like to change.

Also, with the patches for the mentioned issues (new hooks for example) it might be good to set the version to 1.5.0. Let me know what you think.

**Comment by JoryHogeveen** [comment]
@cleverness 
I think that's it for now. Only #20 is left.

Let me know about the version and your review.
In the meantime I will test it on various installations. If/when you have the time, please test it as well.

**Comment by cleverness** [comment]
Yes, definitely a version increase to 1.5.0. This is so awesome! I put it on my schedule to find some time to test it tomorrow.

**Comment by JoryHogeveen** [comment]
Great! And now I really need to address client work haha.
The plugin needed some refactoring to properly build unit tests and address some of the issues I found on wp.org.
Patched it to v1.5.0-rc1 for testing. We'll see what pops up tomorrow!
I'll add some manual tests to the todolist. In the future I can try to build some of those into unit tests.

File: includes/widget-css-classes.class.php [hunk 1]
```diff
@@ -15,94 +15,287 @@
...
-					$instance['ids'] = null;
+			if ( WCSSC_Lib::get_settings( 'show_id' ) ) {
+				if ( $access_id ) {
+					$field = self::do_id_field( $widget, $instance );
+				} else {
+					$field = self::do_hidden( $widget->get_field_name( 'ids' ), $instance['ids'] );
 				}
-				$fields .= "\t<p><label for='widget-{$widget->id_base}-{$widget->number}-ids'>" . apply_filters( 'widget_css_classes_id', esc_html__( 'CSS ID', 'widget-css-classes' ) ) . ":</label>
-				<input type='text' name='widget-{$widget->id_base}[{$widget->number}][ids]' id='widget-{$widget->id_base}-{$widget->number}-ids' value='{$instance['ids']}' class='widefat' /></p>\n";
+				$fields .= "\t" . $field . "\n";
```
Comment (JoryHogeveen) [inline includes/widget-css-classes.class.php]: @cleverness 
I was wondering, why do you add `\t` and `\n` everywhere? Did anyone file some sort of bug report for this?
From what I know this is not needed at all. Let me know!

File: includes/widget-css-classes.class.php [hunk 2]
```diff
@@ -152,114 +349,233 @@ public static function add_widget_classes( $params ) {
...
+		// Add first, last, even, and odd classes.
+		if ( WCSSC_Lib::get_settings( 'show_number' ) ||
+		     WCSSC_Lib::get_settings( 'show_location' ) ||
+		     WCSSC_Lib::get_settings( 'show_evenodd' )
+		) {
 
-		// Add first, last, even, and odd classes
-		if ( 1 === (int) $widget_css_classes['show_number'] || 1 === (int) $widget_css_classes['show_location'] || 1 === (int) $widget_css_classes['show_evenodd'] ) {
 			if ( ! $widget_number ) {
 				$widget_number = array();
```
Comment (JoryHogeveen) [inline includes/widget-css-classes.class.php]: @cleverness 
At the start of this method `$widget_number` is declared as a global variable. Is there a specific reason to make it a global variable? It doesn't look like it's supposed to be modified externally so maybe it's better to declare it as a static variable inside this method.
Let me know!

File: includes/widget-css-classes.class.php (hunk 1 above)
Comment (joseken) [inline includes/widget-css-classes.class.php]: Hmm, they were in the first build in 2012 so it couldn't have been from a bug report. It's not something I would have usually added, so I'm not sure why I did that. They can go though.

File: includes/widget-css-classes.class.php (hunk 2 above)
Comment (cleverness) [inline includes/widget-css-classes.class.php]: I don't think there is a reason it needs to global, but again, that was in the initial build in 2012 I did so the details on why are pretty fuzzy.

**Comment by cleverness** [comment]
I did some quick basic testing from just the previous version to this. I didn't run into any issues.

File: includes/widget-css-classes.class.php (hunk 2 above)
Comment (JoryHogeveen) [inline includes/widget-css-classes.class.php]: Fixed!

File: includes/widget-css-classes.class.php (hunk 1 above)
Comment (JoryHogeveen) [inline includes/widget-css-classes.class.php]: No problemo and fixed!

**Comment by JoryHogeveen** [comment]
Alrighty, I'll make some time today or this weekend to test some more and if that's all good that I think it's good to go!

**Comment by JoryHogeveen** [comment]
@cleverness 
If/when you have the time, can you test again? Also frontend display etc?
I've added another new option "remove duplicate classes" and enhanced the code a bit more.
All tests in the tasklist are done and checked in two of my test sites but an extra pair of eyes is always good!

I'm going to wrap it up for this release, other stuff will need to go in a next one. If you can think of anything I should check/fix/change for 1.5 let me know!

**Pull request merged**